aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+
+logging.basicConfig(level=logging.INFO)
+
+# fix this later to be a proper configured logger
+logger = logging
+
+GCP_PROJECT = 'billing-admin-290403'
+
+GCP_BILLING_BQ_TABLE = (
+    f'{GCP_PROJECT}.billing.gcp_billing_export_v1_01D012_20A6A2_CBD343'
+)
+GCP_AGGREGATE_DEST_TABLE = os.getenv(
+    'GCP_AGGREGATE_DEST_TABLE', 'billing-admin-290403.billing_aggregate.aggregate'
+)
+
+assert GCP_AGGREGATE_DEST_TABLE
+logger.info(f'GCP_AGGREGATE_DEST_TABLE: {GCP_AGGREGATE_DEST_TABLE}')
+
+IS_PRODUCTION = os.getenv('PRODUCTION') in ('1', 'true', 'yes')
+
+# 10% overhead
+HAIL_SERVICE_FEE = 0.05
+ANALYSIS_RUNNER_PROJECT_ID = 'analysis-runner'
+DEFAULT_RANGE_INTERVAL = timedelta(days=2)
+
+
+HAIL_BASE = 'https://batch.hail.populationgenomics.org.au'
+HAIL_UI_URL = HAIL_BASE + '/batches/{batch_id}'
+HAIL_BATCHES_API = HAIL_BASE + '/api/v1alpha/batches'
+HAIL_JOBS_API = HAIL_BASE + '/api/v1alpha/batches/{batch_id}/jobs/resources'
+
+HAIL_PROJECT_FIELD = {
+    'id': 'hail-295901',
+    'number': '805950571114',
+    'name': 'hail-295901',
+    'labels': [],
+    'ancestry_numbers': '/648561325637/',
+    'ancestors': [
+        {
+            'resource_name': 'projects/805950571114',
+            'display_name': 'hail-295901',
+        },
+        {
+            'resource_name': 'organizations/648561325637',
+            'display_name': 'populationgenomics.org.au',
+        },
+    ],
+}
+
+SEQR_PROJECT_FIELD = {
+    'id': 'seqr-308602',
+    'number': '1021400127367',
+    'name': 'seqr-308602',
+    'labels': [],
+    'ancestry_numbers': '/648561325637/',
+    'ancestors': [
+        {
+            'resource_name': 'organizations/648561325637',
+            'display_name': 'populationgenomics.org.au',
+        }
+    ],
+}
+
+
+bigquery_client = bq.Client()
+# pylint: disable=invalid-name
+T = TypeVar('T')
+
+
+async def async_retry_transient_get_json_request(
+    url,
+    errors: Union[Exception, Tuple[Exception, ...]],
+    *args,
+    attempts=5,
+    session=None,
+    **kwargs,
+):
+    """
+    Retry a function with exponential backoff.
+    """
+
+    async def inner_block(session):
+        for attempt in range(1, attempts + 1):
+            try:
+                async with session.get(url, *args, **kwargs) as resp:
+                    resp.raise_for_status()
+                    j = await resp.json()
+                    return j
+            # pylint: disable=broad-except
+            except Exception as e:
+                if not isinstance(e, errors):
+                    raise
+                if attempt == attempts:
+                    raise
+
+            t = 2**attempt
+            logger.warning(f'Backing off {t} seconds for {url}')
+            asyncio.sleep(t)
+
+    if session:
+        return await inner_block(session)
+
+    async with aiohttp.ClientSession() as session2:
+        return await inner_block(session2)
+
+
+def chunk(iterable: Sequence[T], chunk_size=50) -> Iterator[Sequence[T]]:
```

Comment:
Removing the default here might force other functions to be clearer about chunk size

Reply:
Done